src-self-hosted/Module.zig

```diff
@@ -991,14 +991,22 @@ pub fn performAllTheWork(self: *Module) error{OutOfMemory}!void {
                     error.AnalysisFail => {
                         decl.analysis = .dependency_failure;
                     },
+                    error.CGenFailure => {
+                        // Error is handled by CBE, don't try adding it again
+                    },
                     else => {
                         try self.failed_decls.ensureCapacity(self.gpa, self.failed_decls.items().len + 1);
-                        self.failed_decls.putAssumeCapacityNoClobber(decl, try ErrorMsg.create(
-                            self.gpa,
-                            decl.src(),
-                            "unable to codegen: {}",
-                            .{@errorName(err)},
-                        ));
+                        const result = self.failed_decls.getOrPutAssumeCapacity(decl);
+                        if (result.found_existing) {
+                            std.debug.panic("Internal error: attempted to override error '{}' with 'unable to codegen: {}'", .{ result.entry.value.msg, @errorName(err) });
+                        } else {
+                            result.entry.value = try ErrorMsg.create(
+                                self.gpa,
+                                decl.src(),
+                                "unable to codegen: {}",
+                                .{@errorName(err)},
+                            );
+                        }
```

Comment:
Can you explain the reasoning here to change this from the simpler `putAssumeCapacityNoClobber` ?

Reply:
I figured the tiny cost of checking - which only occurs on error paths anyways - is worth it since it helps catch mistakes. You'll probably notice that there's another prong for CGen failures, which does nothing, as CGen errors are added within a deeper scope; this is how I caught what the issue was - essentially, this is to detect if we already added an error for the failed_decl, or if we somehow triggered a path in incremental compilation that doesn't properly remove the failure before updating a decl, etc.

Reply:
OK :+1: